cmd/object-handlers-common.go

```diff
@@ -264,3 +266,69 @@ func setPutObjHeaders(w http.ResponseWriter, objInfo ObjectInfo, delete bool) {
 		}
 	}
 }
+
+// ilmLimitNoncurrentVersions limits the number of noncurrent versions of an
+// object based on lifecycle rules configured on the bucket.
+func ilmLimitNoncurrentVersions(ctx context.Context, objectAPI ObjectLayer, bucket, object string) (err error) {
+	var lc *lifecycle.Lifecycle
+	if lc, err = globalLifecycleSys.Get(bucket); err != nil {
+		if errors.Is(err, BucketLifecycleNotFound{Bucket: bucket}) {
+			return nil
+		}
+		return err
+	}
+
+	objOpts := lifecycle.ObjectOpts{Name: object}
+	lim := lc.NoncurrentVersionsExpirationLimit(objOpts)
+	if lim == 0 { // no limit set
+		return nil
+	}
+
+	var (
+		noncurrentVersions []ObjectToDelete
+		info               ListObjectVersionsInfo
+		keep               int
+	)
+	for {
+		info, err = objectAPI.ListObjectVersions(ctx, bucket, object, info.NextMarker, info.NextVersionIDMarker, "", 1000)
+		if err != nil {
```

Comment:
I can add a function that will give you cheap quorum summary of versions in #13573 - could also be good for healing.